src/core/qgsgeometryvalidator.cpp

```diff
@@ -65,8 +65,8 @@ void QgsGeometryValidator::checkRingIntersections( int p0, int i0, const QgsLine
         {
           d = -distLine2Point( ring1XAtj, ring1YAtj, w.perpVector(), sX, sY );
           if ( d > 0 && d < w.length() &&
-               ring0[i + 1] != ring1[j + 1] && ring0[i + 1] != ring1[j] &&
-               ring0[i + 0] != ring1[j + 1] && ring0[i + 0] != ring1[j] )
+               ring0->pointN( i + 1 ) != ring1->pointN( j + 1 ) && ring1->pointN( i + 1 ) != ring1->pointN( j ) &&
```

Comment:
Seems like you've mixed the rings. Ring0's counter is i and ring1's counter is j

Reply:
oups... you are right, it is a mess!

Reply:
fixed